src/vctrs.h

```diff
@@ -17,6 +35,9 @@ bool is_scalar(SEXP x);
 int equal_object(SEXP x, SEXP y, bool na_equal);
 bool equal_names(SEXP x, SEXP y);
 
+// Caller must ensure proper SEXPTYPEs and lengths for `x` and `y`.
```

Comment:
I still don't quite get what you mean by this. Do you mean that `x` and `y` must have the same `SEXPTYPE`? And that `i` and `j` must be between 0 and `length(x) - 1`/`length(y) - 1`?

Reply:
Exactly. How about:

```c
// Callers must ensure identical SEXPTYPEs for `x` and `y` and lengths
// greater than `i` and `j` respectively. These functions are meant to
// be used in loops which must check bounds.
```

Reply:
I like that! I'd reword slightly:

These functions are meant to be used in loops so it is the callers responsibility to ensure that:
* `x` and `y` have identical SEXTYPES
* `i` is a valid index into `x`, and `j` is a valid index into `y`.
The behaviour is undefined if these conditions are not true.

Reply:
Thanks, done!